src/svd/_bindings.py

```diff
@@ -0,0 +1,306 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+"""
+Various internal functionality used by the bindings module.
+"""
+
+from __future__ import annotations
+
+import enum
+import inspect
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from lxml import objectify
+
+
+class CaseInsensitiveStrEnum(enum.Enum):
+    """String enum class that can be constructed from a case-insensitive string."""
+
+    @classmethod
+    def from_str(cls, value: str) -> CaseInsensitiveStrEnum:
+        """Construct the enum from a case-insensitive string."""
+        value_lower = value.lower()
+        for member in cls:
+            if member.value.lower() == value_lower:
+                return member
+        raise ValueError(
+            f"Class {cls.__qualname__} has no member corresponding to '{value}'"
+        )
+
+
+def to_int(number: str) -> int:
+    """
+    Convert a string representation of an integer following the SVD format to its corresponding
+    integer representation.
+
+    :param number: String representation of the integer.
+
+    :return: Decoded integer.
+    """
+
+    if number.startswith("0x"):
+        return int(number, base=16)
+
+    if number.startswith("#"):
+        return int(number[1:], base=2)
+
+    return int(number)
+
+
+def to_bool(value: str) -> bool:
+    """
+    Convert a string representation of a boolean following the SVD format to its corresponding
+    boolean representation.
+
+    :param value: String representation of the boolean.
+
+    :return: Decoded boolean.
+    """
+    if value in ("true", "1"):
+        return True
+    if value in ("false", "0"):
+        return False
+    raise ValueError(f"Invalid boolean value: {value}")
+
+
+class SvdIntElement(objectify.IntElement):
+    """
+    Element containing an SVD integer value.
+    This class uses a custom parser to convert the value to an integer.
+    """
+
+    def _init(self) -> None:
+        self._setValueParser(to_int)
+
+
+class _Self:
+    """
+    Sentinel value used for element properties where the result class is equal to the class of the
+    parent object. This is required since self-referential class members are not allowed.
+    """
+
+    ...
+
+
+SELF_CLASS = _Self
+
+
+class _Missing:
+    """Sentinel value used to indicate that a default value is missing."""
+
+    ...
+
+
+MISSING = _Missing
+
+
+class elem:
```

Comment:
I'm not sure how I feel about these class names. Are they intended to look like built-ins for when we're objectifying parts of the SVD XML below, or is there another specific reason?

What do you think about doing something like `XmlElem`, to align with PEP8 and link them directly to XML?

Without renaming the file to something like `xml_bindings.py`, I think it makes these more intuitive at a glance that they aren't some reimplementation of a Python built-in, but something XML-specific. Especially `attr`, as that shortname is easily associated with `getattr` and `setattr`.